Woo Installer: Convert transfer step to use v2 endpoint data-layer

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Reworking transfer step for making use of:
* https://github.com/Automattic/wp-calypso/pull/58709 (merged)
* ~D70860-code~ - software endpoint
* ~D71460-code~ ~D70856-code~ - transfer endpoint
* ~D71447-code~ ~D70867-code~ - transfer endpoint software install job
* Rework error handling to close out #58331

#### Testing instructions

- Start an install flow with a simple site via /woocommerce-installation, it should kick off a transfer and install WooCommerce and Woo Payments. It will also require an upgrade if needed.
- Start an install flow with an Atomic site from /woocommerce-installation, it should install WooCommerce and Woo Payments.

Related to #58227

allilevine commented on client/state/atomic-transfer-with-plugin/selectors/index.ts:
```diff
@@ -1 +1,10 @@
+import type { AppState } from 'calypso/types';
+
+export function getAtomicInstallStatus( state: AppState, siteId: number | null ): any {
+	return get( state, [ 'softwareInstall', siteId ?? 0 ], {} );
```
It looks get `get` is commonly used in selectors but we're trying to deprecate lodash in Calypso: p4TIVU-9Bf-p2 Is it worth replacing with [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) or other checks?

lsl replied:
I thought I heard some noise about that. Will update.

lsl commented on client/signup/steps/woocommerce-install/confirm/index.tsx:
```diff
@@ -156,7 +149,7 @@ export default function Confirm( props: WooCommerceInstallProps ): ReactElement
 						<SupportLink />
 						<StyledNextButton
-							disabled={ hasBlockers || ! isDataReady || isAtomicSite }
+							disabled={ hasBlockers || ! isDataReady }
```
~This might need to be reverted, I don't think we display this when the site is already atomic though so maybe not.~
nevermind :smile:

lsl commented on client/signup/steps/woocommerce-install/transfer/install-plugins.tsx:
```diff
@@ -0,0 +1,74 @@
+			setRequestCount( requestCount + 1 );
+			dispatch( requestAtomicSoftwareStatus( siteId, 'woo-on-plans' ) );
+		},
+		requestCount > 5 || softwareStatus.applied ? null : 3000
```
```suggestion
		softwareStatus.applied ? null : 1000
```

debugging

lsl commented on client/signup/steps/woocommerce-install/transfer/install-plugins.tsx:
```diff
@@ -0,0 +1,74 @@
+		}
+	}, [ siteId, goToStep, softwareStatus, wcAdmin ] );
+
+	// todo: Need error handling on these requests
```
^

lsl commented on client/signup/steps/woocommerce-install/transfer/index.tsx:
```diff
@@ -153,19 +31,8 @@ export default function Transfer( props: WooCommerceInstallProps ): ReactElement
-							style={ { '--progress': simulatedProgress } as React.CSSProperties }
-						/>
-					</div>
+					{ isAtomic && <InstallPlugins goToStep={ goToStep } /> }
+					{ ! isAtomic && <TransferSite goToStep={ goToStep } /> }
```
It'd be great if we could combine these two flows cleanly again but I'm not sure we can, maybe using some kind of effect hook to close over the complexity?

lsl replied:
Latest version of this is quite a bit cleaner, at this point the separation helps simplify things.

lsl commented on client/signup/steps/woocommerce-install/transfer/transfer-site.tsx:
```diff
@@ -0,0 +1,96 @@
+	return (
+		<>
+			{ transferFailed && <Error message={ transferStatus || '' } /> }
+			{ ! transferFailed && <Progress progress={ progress } /> }
```
This error handling is still suss, need to confirm what errors look like from this endpoint and how the component behaves. I've put code in that sort of works like I expect it too but haven't checked yet.

lsl commented:
Happy path is now pretty stable for both existing atomic sites and non atomic sites.

You can test by starting an install flow with a simple site, it will kick off a transfer (which will complete) but until D70856-code is merged it won't install software. The transfer component will poll until both transfer is complete and software is applied. 

Since D70860-code is merged, you can test the existing atomic site flow in another tab, it will auto confirm, "re-install" the software set, and redirect to wc-admin.

At the same time, your original tab should complete and redirect to wc-admin.

Once D70856-code testing each flow individually will be possible.

Error handling needs confirming end to end, both transfer failures and http request failures need checking how they're handled.

I think we should just rely on errorNotice for request failures, that pops up a notice bar in the top right of the page. For transfer failures we should set an error state and show it in place of the progress bar.

Code is sort of set up for all this already but I think it just needs testing / working out any bugs. (We can/should do all this in a follow up)

retrofox commented:
![image](https://user-images.githubusercontent.com/77539/145869762-4922a5ad-eb4e-497f-acc9-38fff37da586.png)

Tested the whole process, according to PR instructions and it works as expected.

lsl committed:
Woo Installer: Convert transfer step to use v2 endpoint data-layer

lsl committed:
Dep not required

lsl committed:
Make title optional

The title shown on the install flow was kind of weird as the loading
ellipsis matches the prelim loading checks pre-confirm but then the
title shows up half way through "loading" during the install phase.

lsl committed:
Bugfix progress bar progress

lsl committed:
Transfer style tidy up

lsl committed:
Go back to progress bar + fix smoothing

lsl committed:
Tidy up handling of software status

lsl committed:
dl: fix reducers

lsl committed:
softwareApplied

retrofox committed:
woop: update support link

retrofox committed:
woop: Don't consider Atomic site an error

retrofox committed:
woop: minod comment change

retrofox committed:
woop: enable flow from landing for Atomic sites too

retrofox reviewed (approved):
LGTM. Although it shows an error when installing software when it's an atomic site, it works as expected. We can iterate later in follow-up PRs .

allilevine reviewed (approved):
This is working great now that the endpoint changes are merged. I updated the testing instructions.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7041289

Hi @lsl, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `We've hit a snag`
- `It looks like something went wrong while setting up your store. If this is unexpected, please contact support so that we can help you out.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7041289).
